Fix #2266: [RTL] High-fi Admin and User PinVerification Screen

veena14cs opened the pull request:
## Explanation
This PR fixes #2266 . This PR implements new design for Admin and User Pin Verification
Older implementation of Pin Verification UI was using PinView library which was not supporting RTL. To fix this we removed this library and created new UI as per the new mocks.

## Old Screenshot LTR and RTL
<img width="144" alt="Screenshot_1629793897" src="https://user-images.githubusercontent.com/16301028/130584493-283bb195-afe1-40c5-ad6c-43c02fd2b7c6.png">........<img width="144" alt="Screenshot_1629793938" src="https://user-images.githubusercontent.com/16301028/130584577-191b5ffe-45a0-4002-9ccc-a64a49aee8e1.png">
<img width="256" alt="Screenshot_1629793905" src="https://user-images.githubusercontent.com/16301028/130584597-f33ad0a9-b069-4792-9bd4-ba2bff02df71.png">........<img width="256" alt="Screenshot_1629793945" src="https://user-images.githubusercontent.com/16301028/130584637-44dcddaf-4894-47ad-9cde-99f12bf0e395.png">




## New Screenshot LTR and RTL
<img width="256" alt="Screenshot_1633506133" src="https://user-images.githubusercontent.com/16301028/136087744-b3dba4fa-03c5-40b5-beed-70dde4473e33.png">...<img width="256" alt="Screenshot_1633506065" src="https://user-images.githubusercontent.com/16301028/136087827-a92f0ade-50d1-44ee-a1c3-78483949c24e.png">

<img width="144" alt="Screenshot_1633506118" src="https://user-images.githubusercontent.com/16301028/136087777-75161bb1-e385-4051-8c43-ffa3db9f836c.png">.....<img width="144" alt="Screenshot_1633506083" src="https://user-images.githubusercontent.com/16301028/136087906-44647741-13bc-4c33-a931-9282077a73d2.png">


Tablet


<img width="160" alt="Screenshot_1633504178" src="https://user-images.githubusercontent.com/16301028/136083376-f5864ea7-b8ce-48c9-b6e6-a1aa52bc7b8d.png">.....<img width="160" alt="Screenshot_1633504220" src="https://user-images.githubusercontent.com/16301028/136083423-58e2b3d2-642f-4070-ac35-712954c94958.png">

<img width="256" alt="Screenshot_1633504169" src="https://user-images.githubusercontent.com/16301028/136083500-75b838af-99c1-4fa1-bbdf-8e2a236b9f74.png">.....<img width="256" alt="Screenshot_1633504231" src="https://user-images.githubusercontent.com/16301028/136083584-08162b4b-9697-45e6-b51a-a89cb1804c74.png">


## Mock links
Mobile Portrait : https://xd.adobe.com/view/3dca36c2-5115-419c-b25e-0f10526b077c-6899/screen/4d274b3e-c27b-438a-8c74-c0ff16134080/
Mobile Landscape : https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/screen/327723da-a1e8-40e4-aace-9c30d663b20b/
Tablet Portrait : https://xd.adobe.com/view/d405de00-a871-4f0f-73a0-f8acef30349b-a234/screen/8de829a5-fc1c-4083-b510-44c1387ed9c2/
Tablet Landscape : https://xd.adobe.com/view/d405de00-a871-4f0f-73a0-f8acef30349b-a234/screen/0648610b-7d70-403c-a42e-a206be75e85f/

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

veena14cs committed:
Update selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update return_to_topic_button_item.xml

veena14cs committed:
Update return_to_topic_button_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Changed Pin password UI

veena14cs committed:
Fixed margins

veena14cs committed:
removed unwanted code

veena14cs committed:
fixed testcases

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs committed:
fixed lint

veena14cs committed:
Delete PinViewBindingAdapters.java

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

rt4914 commented on app/build.gradle:
```diff
@@ -110,7 +110,6 @@ dependencies {
       'androidx.multidex:multidex:2.0.1',
       'androidx.recyclerview:recyclerview:1.0.0',
       'androidx.work:work-runtime-ktx:2.4.0',
-      'io.github.chaosleung:pinview:1.4.4',
```
remove this from bazel files too.

veena14cs replied:
Removed library from bazel.

BenHenning replied:
Please also remove from versions.bzl.

veena14cs replied:
Done.

rt4914 reviewed (changes requested):
@veena14cs Currently the mocks are used such that you have kept extreme right and left margin/padding as fixed as per mocks and the content inside is either `wrap` or `match`.

But instead the content should be fixed width and margin/padding should be variable.

For example if you see tablet mocks the content width is of `360dp`, so we should fix that and make margins variable. Similarly the `inputbox + show/hide` is `304dp` in mobile.

Also the background/toolbar is completely different as compared to mocks.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs commented:
> @veena14cs Currently the mocks are used such that you have kept extreme right and left margin/padding as fixed as per mocks and the content inside is either `wrap` or `match`.
> 
> But instead the content should be fixed width and margin/padding should be variable.
> 
> For example if you see tablet mocks the content width is of `360dp`, so we should fix that and make margins variable. Similarly the `inputbox + show/hide` is `304dp` in mobile.

Shouldn't we keep variable width to support smaller devices?

veena14cs committed:
fixed margins

veena14cs committed:
Update pin_password_activity.xml

veena14cs commented:
> > @veena14cs Currently the mocks are used such that you have kept extreme right and left margin/padding as fixed as per mocks and the content inside is either `wrap` or `match`.
> > But instead the content should be fixed width and margin/padding should be variable.
> > For example if you see tablet mocks the content width is of `360dp`, so we should fix that and make margins variable. Similarly the `inputbox + show/hide` is `304dp` in mobile.
> 
> Shouldn't we keep variable width to support smaller devices?

I fixed the changes PTAL. Margins and width I have kept variable size to support all types of screen.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs committed:
Update BUILD.bazel

rt4914 commented on app/src/main/res/layout/pin_password_activity.xml:
```diff
@@ -14,127 +14,186 @@
-        android:textSize="12sp" />
-    </LinearLayout>
+        android:layout_gravity="center"
+        android:paddingTop="42dp">
```
Make this value `40` instead of `42dp`

veena14cs replied:
Done.

rt4914 commented on app/src/main/res/layout/pin_password_activity.xml:
```diff
@@ -14,127 +14,186 @@
+          android:lines="1"
+          android:text="@{String.format(@string/pin_password_hello, viewModel.profile.name)}"
+          android:textColor="@color/oppiaPrimaryText"
+          android:textSize="24sp"
```
For different textviews you can use styles from `styles.xml` file.
For example for this textview you can change it too
```
<TextView
          style="@style/Heading1"
          android:id="@+id/hello_text_view"
          android:fontFamily="sans-serif-medium"
          android:text="@{String.format(@string/pin_password_hello, viewModel.profile.name)}"
          app:layout_constraintEnd_toEndOf="parent"
          app:layout_constraintHorizontal_bias="0.0"
          app:layout_constraintStart_toStartOf="parent"
          app:layout_constraintTop_toTopOf="parent" />
```

rt4914 replied:
Ditto, applies to other textviews in all files.

veena14cs replied:
Done.

rt4914 commented on app/src/main/res/layout/pin_password_activity.xml:
```diff
@@ -14,127 +14,186 @@
+          android:layout_width="wrap_content"
+          android:layout_height="wrap_content"
+          android:fontFamily="sans-serif-medium"
+          android:lines="1"
```
Why does it need to be in `1` line only? I think it should be completely fine to expand it to mutliple lines as we do have space on this screen.

veena14cs replied:
removed this line.

rt4914 reviewed (changes requested):
For earlier [comment](https://github.com/oppia/oppia-android/pull/3684#pullrequestreview-732029890) the background/toolbar color changes have been missed. The entire screen is of yellow/cream color instead of blue color in mocks.

Also you can merge `pin_password_activity.xml` files by mentioning all values in dimens.xml file. 
For this follow these steps and it would become much easy:
1. On [text-compare](https://text-compare.com/) on left paste content from `mobile-portrait` file and on right paste from `mobile-landscape` and now for all difference create values in `dimens.xml` file and its other qualifier files.
2. Apply this same process for `mobile-portrait` vs `tablet-portrait` and `mobile-portrait` vs `tablet-portrait` too.

veena14cs committed:
fixed changes

veena14cs committed:
Fixed background color

veena14cs commented:
> For earlier [comment](https://github.com/oppia/oppia-android/pull/3684#pullrequestreview-732029890) the background/toolbar color changes have been missed. The entire screen is of yellow/cream color instead of blue color in mocks.
> 
> Also you can merge `pin_password_activity.xml` files by mentioning all values in dimens.xml file.
> For this follow these steps and it would become much easy:
> 
> 1. On [text-compare](https://text-compare.com/) on left paste content from `mobile-portrait` file and on right paste from `mobile-landscape` and now for all difference create values in `dimens.xml` file and its other qualifier files.
> 2. Apply this same process for `mobile-portrait` vs `tablet-portrait` and `mobile-portrait` vs `tablet-portrait` too.

I have fixed changes and updated screenshots in PR description.
 I think we cannot merge  `pin_password_activity.xml` files as mobile landscape screen is different. Rest all files are same.

veena14cs committed:
Update pin_password_activity.xml

rt4914 commented:
> > For earlier [comment](https://github.com/oppia/oppia-android/pull/3684#pullrequestreview-732029890) the background/toolbar color changes have been missed. The entire screen is of yellow/cream color instead of blue color in mocks.
> > Also you can merge `pin_password_activity.xml` files by mentioning all values in dimens.xml file.
> > For this follow these steps and it would become much easy:
> > 
> > 1. On [text-compare](https://text-compare.com/) on left paste content from `mobile-portrait` file and on right paste from `mobile-landscape` and now for all difference create values in `dimens.xml` file and its other qualifier files.
> > 2. Apply this same process for `mobile-portrait` vs `tablet-portrait` and `mobile-portrait` vs `tablet-portrait` too.
> 
> I have fixed changes and updated screenshots in PR description.
> I think we cannot merge `pin_password_activity.xml` files as mobile landscape screen is different. Rest all files are same.


@veena14cs Following changes are needed:
1. In `PinPasswordActivityPresenter` change the status bar color so that it can match mock. It might be doable by removing this line : `StatusBarColor.statusBarColorUpdate(R.color.pinInputStatusBar, activity, true)`
2. The shadow below toolbars is not visible. Check `add_profile_activity.xml` line `247` for reference.
3. There is one bug in this. Type 2 digits, click on show/hide and see how cursor shifts to start. This needs to be fixed.
4. For merging all xml files. Actually it is possible because the difference between mobile-portrait and mobile-landscape file is one line only which is `app:layout_constraintHorizontal_bias="0.0"` is part of `pin_password_input_pin` in portrait and it not available in landscape. This can be fixed by introducing `app:layout_constraintHorizontal_bias="0.5"` landscape also. Now we need to shift these values to `dimens` file which can be done with this code https://stackoverflow.com/a/8780360

veena14cs committed:
fixed bug and merged files

veena14cs committed:
Update PinPasswordActivityPresenter.kt

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
resolved conflicts

veena14cs commented:
> @veena14cs Following changes are needed:
> 
> 1. In `PinPasswordActivityPresenter` change the status bar color so that it can match mock. It might be doable by removing this line : `StatusBarColor.statusBarColorUpdate(R.color.pinInputStatusBar, activity, true)`
> 2. The shadow below toolbars is not visible. Check `add_profile_activity.xml` line `247` for reference.
> 3. There is one bug in this. Type 2 digits, click on show/hide and see how cursor shifts to start. This needs to be fixed.
> 4. For merging all xml files. Actually it is possible because the difference between mobile-portrait and mobile-landscape file is one line only which is `app:layout_constraintHorizontal_bias="0.0"` is part of `pin_password_input_pin` in portrait and it not available in landscape. This can be fixed by introducing `app:layout_constraintHorizontal_bias="0.5"` landscape also. Now we need to shift these values to `dimens` file which can be done with this code https://stackoverflow.com/a/8780360

@rt4914 thanks for the above reference it was really helpful. I have done changes to the list mentioned above.
1) Fixed statusbar.
2) Added shadow below toolbar.
3) Fixed bug of Edit text cursor on show/hide click event.
4) Merged `pin_password_activity.xml`

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs committed:
Update BUILD.bazel

veena14cs committed:
Update pin_password_activity.xml

veena14cs committed:
fixed dimensions

veena14cs committed:
removed Pinview library

veena14cs committed:
reverted changes

veena14cs committed:
Update pin_password_activity.xml

rt4914 reviewed (approved):
LGTM, thanks.

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -265,7 +265,9 @@
   <string name="pin_password_admin_enter">Please enter your Administrator PIN.</string>
   <string name="pin_password_user_enter">Please enter your PIN.</string>
-  <string name="pin_password_forgot_pin">I forgot my pin.</string>
+  <string name="pin_password_admin_pin_enter">Administrator’s 5-Digit PIN.</string>
+  <string name="pin_password_user_pin_enter">User’s 3-Digit PIN.</string>
+  <string name="pin_password_forgot_pin">I forgot my pin</string>
```
These need more context I think on where the text shows up. For example, the first text could be:

``input_pin_password_as_admin``

Please update all new string names accordingly to have more context.

veena14cs replied:
Updated strings.

BenHenning reviewed (changes requested):
@veena14cs had a few comments. Also, could you please add screenshots for the 'before' screen to show more explicitly what changed?

Thanks.

veena14cs committed:
removed library from versions.bzl and fixed strings file

veena14cs commented:
> @veena14cs had a few comments. Also, could you please add screenshots for the 'before' screen to show more explicitly what changed?
> 

Added before and after screenshots.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs committed:
Update pin_password_activity.xml

veena14cs committed:
Update maven_dependencies.textproto

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs committed:
reverted removal of library

veena14cs committed:
Update maven_install.json

veena14cs committed:
reverted file

veena14cs committed:
Create maven_install.json

veena14cs committed:
Merge branch 'hi-fi-pinview' of https://github.com/veena14cs/oppia-android into hi-fi-pinview

veena14cs committed:
revert maven files

veena14cs committed:
Delete maven_install.json

veena14cs commented:
@prayutsu as we discussed in hangouts meeting and tried to solve maven update error. It wasn't working on Windows and as its on priority I would assign it to you to remove third party library and run this script using bazel command to update maven_install.json file.

prayutsu committed:
Remove unwanted dependency

prayutsu commented:
> @prayutsu as we discussed in hangouts meeting and tried to solve maven update error. It wasn't working on Windows and as its on priority I would assign it to you to remove third party library and run this script using bazel command to update maven_install.json file.

@veena14cs Done.

veena14cs committed:
Trigger Build

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Update PinPasswordActivityTest.kt

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
resolved conflicts

veena14cs committed:
updated with latest develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Resolved conflict

BenHenning commented on .idea/misc.xml:
```diff
@@ -1,9 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" project-jdk-name="1.8" project-jdk-type="JavaSDK">
+  <component name="ProjectRootManager" version="2" languageLevel="JDK_1_8" default="true" project-jdk-name="1.8" project-jdk-type="JavaSDK">
```
Unrelated change--please revert.

veena14cs replied:
Done.

BenHenning reviewed (changes requested):
Thanks for adding the old screenshots @veena14cs.

Reviewed my codeowners files (except strings), but before I go further I have some fundamental concerns with this. @rt4914 we discussed that we should prefer keeping the existing user experience. Was a decision made somewhere to change the UX--did I possibly miss that? Also, was Chantel included in that decision & is the rationale documented? If so, could we please capture that in this PR. If not, we should make sure UX weighs in on this since it's a major user experience change.

rt4914 commented:
> Thanks for adding the old screenshots @veena14cs.
> 
> Reviewed my codeowners files (except strings), but before I go further I have some fundamental concerns with this. @rt4914 we discussed that we should prefer keeping the existing user experience. Was a decision made somewhere to change the UX--did I possibly miss that? Also, was Chantel included in that decision & is the rationale documented? If so, could we please capture that in this PR. If not, we should make sure UX weighs in on this since it's a major user experience change.

@BenHenning Yes we did finalise on changing the UX. @seanlip  and @mschanteltc are aware of this and the mocks were provided by Chantel. The main rationale was that actually there is no big preference from Chantel's side and she thinks either UI is fine and considering we were have a lot of technical difficulties fixing the previous UI we finally changed it. Also I have added you to that email thread today so you can have a look at the entire conversation.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs committed:
Reverted misc.xml file

veena14cs commented:
> > Thanks for adding the old screenshots @veena14cs.
> > Reviewed my codeowners files (except strings), but before I go further I have some fundamental concerns with this. @rt4914 we discussed that we should prefer keeping the existing user experience. Was a decision made somewhere to change the UX--did I possibly miss that? Also, was Chantel included in that decision & is the rationale documented? If so, could we please capture that in this PR. If not, we should make sure UX weighs in on this since it's a major user experience change.
> 
> @BenHenning Yes we did finalise on changing the UX. @seanlip and @mschanteltc are aware of this and the mocks were provided by Chantel. The main rationale was that actually there is no big preference from Chantel's side and she thinks either UI is fine and considering we were have a lot of technical difficulties fixing the previous UI we finally changed it. Also I have added you to that email thread today so you can have a look at the entire conversation.


@BenHenning does this help in clarifying the changes? If yes can you please review this PR.

BenHenning commented on app/src/main/res/values/strings.xml:
```diff
@@ -275,6 +277,7 @@
   <string name="password_shown_icon">Password shown icon</string>
   <string name="password_hidden_icon">Password hidden icon</string>
   <string name="enter_your_pin">Enter your PIN</string>
+  <string name="pin_password_toolbar">Enter PIN</string>
```
Should this be ``pin_password_toolbar_title``? It's a bit unclear on what the "what" is corresponding to this string.

veena14cs replied:
Changed it to `pin_password_toolbar_title`.

BenHenning reviewed (commented):
Thanks @rt4914 and @veena14cs. Yes, this helps clarify. Just had 1 comment, otherwise the PR LGTM. PTAL.

veena14cs committed:
renamed label

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Resolved conflicts

BenHenning reviewed (changes requested):
Thanks @veena14cs. This LGTM from a code perspective (per my codeowner files), but it looks like CI is failing. In the very least, new dependencies need to be pinned for Maven. Can you please fix the CI failures & assign back?

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs commented:
> Thanks @veena14cs. This LGTM from a code perspective (per my codeowner files), but it looks like CI is failing. In the very least, new dependencies need to be pinned for Maven. Can you please fix the CI failures & assign back?

Thanks @BenHenning. To fix CI failures I need to run bazel in order to update Maven. @rt4914 can you please do it from your system and push the code as it is blocking me from submitting this PR.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
resolved conflicts

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs committed:
reverted bazel files

veena14cs committed:
Update image_region_selection_interaction_item.xml

veena14cs committed:
Update image_region_selection_interaction_item.xml

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

FareesHussain committed:
removed io_github_chaosleung_pinview from version.bzl

FareesHussain committed:
removed dependency from textproto

FareesHussain committed:
updated textproto

rt4914 commented:
@BenHenning @veena14cs 
I am trying to follow these steps so as to fix the bazel issues.
1. Remove `pin-view` version from `versions.bzl` file.
2. Run `REPIN=1 bazel run @unpinned_maven//:pin` in terminal (It was successful)
3. Run `bazel run //scripts:generate_maven_dependencies_list -- $(pwd) third_party/maven_install.json scripts/assets/maven_dependencies.textproto scripts/assets/maven_dependencies.pb` in terminal.

I am getting following error:
<img width="500" alt="Screenshot 2021-09-25 at 2 42 51 AM" src="https://user-images.githubusercontent.com/9396084/134740391-0920be1e-ec57-43c9-9c10-25b129a420ae.png">


After this I checked out to `develop` too and followed step 2 and 3 from above and I am still getting that same error.

@BenHenning Any idea what I am doing wrong?

veena14cs committed:
Update maven_dependencies.textproto

veena14cs committed:
reverted files

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/PinPasswordActivityPresenter.kt:
```diff
@@ -52,50 +49,66 @@ class PinPasswordActivityPresenter @Inject constructor(
       viewModel = pinViewModel
     }
 
+    binding.pinPassordToolbar.setNavigationOnClickListener {
+      (activity as PinPasswordActivity).finish()
+    }
```
Why is this needed now? It doesn't seem like we did it this way before.

veena14cs replied:
In older UI we were not having toolbar. In new version of Pin VIew UI toolbar is been introduced.

BenHenning reviewed (commented):
One comment @veena14cs, otherwise this is also a reminder to re-remove the pinview dependency from versions.bzl once we figure out what's going wrong with pinning.

> @BenHenning @veena14cs
> I am trying to follow these steps so as to fix the bazel issues.
> 
> 1. Remove `pin-view` version from `versions.bzl` file.
> 2. Run `REPIN=1 bazel run @unpinned_maven//:pin` in terminal (It was successful)
> 3. Run `bazel run //scripts:generate_maven_dependencies_list -- $(pwd) third_party/maven_install.json scripts/assets/maven_dependencies.textproto scripts/assets/maven_dependencies.pb` in terminal.
> 
> I am getting following error:
> <img alt="Screenshot 2021-09-25 at 2 42 51 AM" width="500" src="https://user-images.githubusercontent.com/9396084/134740391-0920be1e-ec57-43c9-9c10-25b129a420ae.png">
> 
> After this I checked out to `develop` too and followed step 2 and 3 from above and I am still getting that same error.
> 
> @BenHenning Any idea what I am doing wrong?

@rt4914 are you able to build with Bazel locally? I.e. does ``bazel build //:oppia`` work? It looks like you have an Android SDK incompatibility with the Bazel environment, though I'm not clear yet on what the actual issue is.

BenHenning commented:
Note that I'm building locally & will create a Gist with the file updates needed.

BenHenning commented:
@veena14cs PTAL at https://gist.github.com/BenHenning/9e541aa731d5af2379d766a117ad3929/revisions & particularly the latest revision https://gist.github.com/BenHenning/9e541aa731d5af2379d766a117ad3929. That should contain the changes you need to remove the dependency & pass CI.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs committed:
removed library and manually updated dependencies

veena14cs committed:
Update maven_install.json

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs commented:
> @veena14cs PTAL at https://gist.github.com/BenHenning/9e541aa731d5af2379d766a117ad3929/revisions & particularly the latest revision https://gist.github.com/BenHenning/9e541aa731d5af2379d766a117ad3929. That should contain the changes you need to remove the dependency & pass CI.

@BenHenning I have updated files and also CI checks are passing now. PTAL

BenHenning commented on scripts/assets/maven_dependencies.textproto:
```diff
@@ -872,4 +861,4 @@ maven_dependency {
     }
   }
-}
+}
```
Please add EOF newline.

veena14cs replied:
Done.

BenHenning commented on third_party/versions.bzl:
```diff
@@ -132,7 +131,7 @@ HTTP_DEPENDENCY_VERSIONS = {
     },
     "rules_proto": {
-        "sha": "e0cab008a9cdc2400a1d6572167bf9c5afc72e19ee2b862d18581051efab42c9",
+        "sha": "602e7161d9195e50246177e7c55b2f39950a9cf7366f74ed5f22fd45750cd208",
```
Why this change? It seems unrelated--I suggest reverting this part.

veena14cs replied:
Reverted.

BenHenning reviewed (commented):
Thanks @veena14cs. LGTM other than the two new comments.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
resolved conflicts

veena14cs committed:
merged layout

veena14cs committed:
Update maven_install.json

veena14cs committed:
Update maven_install.json

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs committed:
fixed code changes

veena14cs committed:
reverted dependency files

veena14cs committed:
reverted library

veena14cs committed:
Update pin_password_activity.xml

veena14cs commented:
@BenHenning all tests are failing in `PinPasswordActivityTest`

<img width="384" alt="Screenshot (13)" src="https://user-images.githubusercontent.com/16301028/135254927-adfd613a-8951-4a5d-96bd-b5af1f947649.png">


I tested on develop too. It fails with the same error.

<img width="384" alt="Screenshot (14)" src="https://user-images.githubusercontent.com/16301028/135255035-6131b950-73f8-444a-8565-2a82c8f7bb67.png">

BenHenning commented:
@veena14cs did you try running that suite first on Robolectric? Per the CI logs, https://gist.github.com/BenHenning/fe5e145b8720e45f7f1afbf61e0af120 seem to be the actual blocking failures. I suggest trying to fix these first then retry the Espresso run.

That being said, the error you're seeing on Espresso shouldn't be happening per the initialization rule. Something definitely seems wrong there.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs commented:
> @veena14cs did you try running that suite first on Robolectric? Per the CI logs, https://gist.github.com/BenHenning/fe5e145b8720e45f7f1afbf61e0af120 seem to be the actual blocking failures. I suggest trying to fix these first then retry the Espresso run.
> 
> That being said, the error you're seeing on Espresso shouldn't be happening per the initialization rule. Something definitely seems wrong there.

@BenHenning these tests were passing earlier and now I have added `land` qualifiers to pass this tests.

veena14cs committed:
fixed tests

veena14cs committed:
Update PinPasswordActivityTest.kt

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/profile/PinPasswordActivityTest.kt:
```diff
@@ -951,6 +953,7 @@ class PinPasswordActivityTest {
   }
 
   @Test
+  @Config(qualifiers = "land-xhdpi")
```
@rt4914 is landscape actually correct to use for these?

@veena14cs this might just be a case where the new UI isn't fitting within the screen on Robolectric & we now need more room in order for the test to pass. Does it pass with something like:

```kotlint
@Config(qualifiers = "w360dp-h760dp-port-xxxhdpi")
```

(I think that roughly corresponds to a Pixel XL)

veena14cs replied:
I did some changes in the UI and now tests passes without this.

BenHenning replied:
@rt4914 can you re-review the layout changes & approve if they look good to you? Test changes seem good to me.

BenHenning replied:
Marking as resolved since Rajat is taking a look.

BenHenning reviewed (commented):
Thanks @veena14cs. The PR looks good, but I'm not sure about the test change. Changing to landscape seems like it might be a slight hack--maybe we can keep it in portrait but increase the size so that no scrolling is needed & all needed elements fit on the screen?

veena14cs commented:
> Thanks @veena14cs. The PR looks good, but I'm not sure about the test change. Changing to landscape seems like it might be a slight hack--maybe we can keep it in portrait but increase the size so that no scrolling is needed & all needed elements fit on the screen?

@BenHenning I have fixed the UI, may be somewhere it got changed while updating it with develop or while merging xml files into one and it went missing. Now, its fixed PTAL.

veena14cs committed:
fixed tests

veena14cs committed:
Update pin_password_activity.xml

BenHenning commented on app/src/main/res/layout/pin_password_activity.xml:
```diff
@@ -14,127 +14,196 @@
+    </FrameLayout>
   </androidx.constraintlayout.widget.ConstraintLayout>
-</layout>
+</layout>
```
Please add EOF newline.

veena14cs replied:
Done.

BenHenning reviewed (commented):
Thanks @veena14cs. PR generally LGTM, but I think Rajat should take one more pass since there seemed to be a lot of layout changes in the last commit.

veena14cs committed:
fixed ui changes

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

rt4914 reviewed (changes requested):
@veena14cs The UI does not look correct.
1. The label `Administrator's 5 Digit PIN` UI is different from that of `admin_auth_activity`.
2. In landscape mode the UI does not look in centre.
3. Toolbar shadow is not correct.
![Screenshot_1633438399](https://user-images.githubusercontent.com/9396084/136027021-8ae0015c-5828-4f0a-883d-c9e1798da9e8.png)


Also, suggest updating the screenshots in PR Description

veena14cs committed:
added EOF

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

veena14cs commented:
> @veena14cs The UI does not look correct.
> 
> 1. The label `Administrator's 5 Digit PIN` UI is different from that of `admin_auth_activity`.
> 2. In landscape mode the UI does not look in centre.
> 3. Toolbar shadow is not correct.
>    ![Screenshot_1633438399](https://user-images.githubusercontent.com/9396084/136027021-8ae0015c-5828-4f0a-883d-c9e1798da9e8.png)
> 
> Also, suggest updating the screenshots in PR Description

@rt4914 I have addressed and updated the changes
1. Changed the UI keeping it consistent with `admin_auth_activity`.
2. Centered UI at the center in landscape mode.
3. Corrected Toolbar shadow in Tablet mode.

Updated screenshots in PR description. PTAL Thanks.

BenHenning reviewed (approved):
Thanks @veena14cs. This LGTM; deferring to Rajat for final approval.

@rt4914 please re-add me if anything changes from the build or testing side.

veena14cs committed:
Fixed hint and landscape ui for tablet

veena14cs committed:
fixed dimens

veena14cs committed:
Fixed dimens

veena14cs committed:
Update pin_password_activity.xml

veena14cs committed:
removed testcase that tests focus of the keyboard by default

veena14cs committed:
reverted default keyboard focus

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview

rt4914 reviewed (approved):
LGTM, thanks.

veena14cs committed:
Merge remote-tracking branch 'upstream/develop' into develop

veena14cs committed:
Merge branch 'develop' into hi-fi-pinview
